### PR TITLE
fix: connected types consolidate type definitions with and without a condition

### PR DESCRIPTION
**Pull request by willvedd**
## Description
Currently, when creating a typesystem from a model, we reduce certain relations' indirections into their terminal relations. This can be used to understand when we can apply "fast-path" optimization for TTU and userset subproblems. Initially, these optimizations only account for rewrites other than union, intersection and exclusion. By excluding these rewrites we can assert that there is always a single terminal relation for a certain combination of user type, object type and relation. 
However, we erroneously "double-count" under during two scenarios:
- multiple type restrictions of a single type – with and without a condition (example 1 below)
- multiple type restrictions of a single type – with multiple conditions (example 2 below)

This presents as a 500 error with a message of `"expected exactly one terminal relation for fast path, received 2"` when running a `Check` where a subproblem incurs one or more described scenarios above.

**Example 1:**
```
model
   schema 1.1

type user
type group
   relations
      define member: [user, user with x_less_than] # Note: `user` with and without condition

type folder
   relations
      define parent: [group]
      define viewer: member from parent

condition x_less_than(x: int) {
   x < 100
}
```
Calling `Check( "user:1", "viewer", "folder:1")`

**Example 2:**
```
model
   schema 1.1

type user
type group
   relations
      define member: [user with x_less_than, user with  x_greater_than] # Note: multiple conditions

type folder
   relations
      define parent: [group]
      define viewer: member from parent

condition x_less_than(x: int) {
   x < 100
}

condition x_greater_than(x: int) {
   x < 100
}
```
Calling `Check( "user:1", "viewer", "folder:1")`

This PR fixes this "double-count" by preventing duplicate terminal relations from being recorded in the typesystem.

## References
Introduced by #1765 

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: assets/tests/abac_tests.yaml [hunk 1]
```diff
@@ -1092,6 +1092,184 @@ tests:
...
+              user: user:with-condition
+              relation: viewer
+              object: document:1
+            errorCode: 2000
+          - tuple:
+              user: user:with-condition
+              relation: viewer
+              object: document:1
+            context:
+              "x": 10
```
Comment (adriantam) [inline assets/tests/abac_tests.yaml:1134]: can we also add the case where x != 10

File: assets/tests/abac_tests.yaml [hunk 2]
```diff
@@ -1092,6 +1092,184 @@ tests:
...
+            expectation: true
+          - tuple:
+              user: user:without-condition
+              relation: viewer
+              object: document:1
+            expectation: true
+          - tuple:
```
Comment (adriantam) [inline assets/tests/abac_tests.yaml:1141]: how is this different than the tuple above?

File: assets/tests/abac_tests.yaml [hunk 3]
```diff
@@ -1092,6 +1092,184 @@ tests:
...
+              object: document:1
+            context:
+              "x": 10
+            expectation: true
+          - tuple:
+              user: user:with-ycond
+              relation: viewer
+              object: document:1
+            context:
+              "y": 10
```
Comment (adriantam) [inline assets/tests/abac_tests.yaml:1230]: Please also add the negative case

File: assets/tests/abac_tests.yaml [hunk 4]
```diff
@@ -1092,6 +1092,184 @@ tests:
...
+              user: user:without-condition
+              relation: viewer
+              object: document:1
+            expectation: true
+          - tuple:
+              user: user:without-condition
+              relation: viewer
+              object: document:1
+            expectation: true
+        listUsersAssertions:
```
Comment (adriantam) [inline assets/tests/abac_tests.yaml:1146]: Do we have list objects assertion?

Comment (willvedd) [reply assets/tests/abac_tests.yaml:1175]: Added for both integration tests.

File: assets/tests/abac_tests.yaml (hunk 2 above)
Comment (willvedd) [reply assets/tests/abac_tests.yaml:1143]: Removed.

File: assets/tests/abac_tests.yaml (hunk 1 above)
Comment (willvedd) [reply assets/tests/abac_tests.yaml:1134]: Added.

**Review by adriantam — APPROVED** [review]

**Review by elbuo8 — APPROVED** [review]

**Pull request merged**